### ISSUE TITLE
fitter + hold doesn't work very well

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#13]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/13) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

When I've tried to use the hold feature of the fitter I have gotten much worse results than without it, even when it seems like it should make sense.  Maybe instead of just not changing a parameter, it should actually call the underlying algorithm with fewer parameters?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Well, I am pretty sure that it actually *does* call the underlying algorithm with fewer parameters, in a (possibly) convoluted way.

Are you certain that you are supplying sensible starting parameters?  The algorithm to guess good parameters from which to search is quite fragile, and I would be very surprised if it works at all when there's a held parameter. You almost certainly have to choose your own good starting parameters when holding.

If you are actually hoping to get a bug fixed, then show some example code and data that makes it fail.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

This code creates data with known parameters, feeds the exactly correct parameters to the fitter as guesses, and plots the results. The only difference between the two plots is that I ask it to hold parameter 2 (aka the energy/pulse_height slope) in the second (aka right) plot.
```
#!python

import mass
import pylab as plt
import numpy as np

fitter = mass.calibration.MnKAlphaFitter()

        # params: a 6-element sequence of [Resolution (fwhm), Pulseheight of the Kalpha1 peak,
        #         energy scale factor (pulseheight/eV), amplitude, background level (per bin),
        #         and background slope (in counts per bin per bin) ]
        #         If params is None, all params will be guessed.
        #         If params is a 6 element list, all elements with value None will be guessed.
params = np.array([1.0,fitter.spect.peak_energy,1.0,100.,0.001,0.001], dtype=np.float)
erange = np.array((fitter.spect.peak_energy-100, fitter.spect.peak_energy+100))
binsize = 1
scale = 100
x = np.arange(erange[0], erange[1], binsize)
y_ideal = fitter.fitfunc(params, x)
y_data = np.random.poisson(fitter.fitfunc(params, x)*scale)

pGuess = [params[0], params[1], params[2], params[3], params[4], params[5]]
plt.figure()
ax = plt.gca()
fitter.fit(y_data, x, params=pGuess, axis=ax)
plt.figure()
ax = plt.gca()
fitter.fit(y_data, x, params=pGuess, hold=[2], axis=ax)
```

![Figure 2.png](https://bitbucket.org/repo/KrBpjo/images/2794272581-Figure%202.png)

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Fix bug in how maximum likelihood fitter behaves when params are held.

Fixes issue #13. If parameters being held were the last N, then bug was
not seen. If they were, say, [2,5] out of 6, though, then certain calcs
were wrong because they assumed that the 2 parameters being held had to
be the LAST two, i.e., [4,5] out of 6.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Thanks, Galen, for making a simple test case. If you run the same code with the new Mass, I think you'll find it works now.
